### DISCUSSION TITLE
[FEATURE REQUEST]:  Extensibility of UDAF

### DISCUSSION THREAD

**Discussion by samueleresca** [dotnet/spark#531] [Q&A] [answered]
Hi, 
I was trying to replicating the following implementation
[awslabs/deequ/StatefulDataType](https://github.com/awslabs/deequ/blob/5be86f479ec9a4132017121839c2baae106cc963/src/main/scala/com/amazon/deequ/analyzers/catalyst/StatefulDataType.scala#L26) using .NET Spark.

The `StatefulDataType` class extends the `UserDefinedAggregateFunction` and it internally uses the `AggregateExpression` to parse a sequence of Columns in the [apply](https://github.com/apache/spark/blob/176b69642e217883be47951654b37c86d7d82eaf/sql/core/src/main/scala/org/apache/spark/sql/expressions/udaf.scala#L151) method.

Are these APIs already supported in .NET spark? Do you have any plan to support them? Is there any alternative way I can implement the `StatefulDataType` class using the already existing .NET Spark API?

Thanks

**Comment by imback82** [comment, marked as answer]
@samueleresca Unfortunately, `UserDefinedAggregateFunction` is not supported in .NET for Apache Spark. If you already have a Scala implementation, you can use `RegisterJavaUDAF` to register it and use it in SQL commands.

Just FYI, `UserDefinedAggregateFunction` will be deprecated in Spark 3.0 in favor of `Aggregator`.

**Comment by AnakinSkyCN** [comment]
waiting for the UserDefinedAggregateFunction, so that u wouldn't have to sort the entire DataFrame just to find a Row that holds the max value of a column.
